### PR TITLE
[#5243] Adding progress reporter to bulk operations

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: https://github.com/Azure/azure-sdk-for-java/issues/5243

Wraps implementation for IProgressReceiver to trigger `reportProgress` on each upload/download step

- adding a simple test and samples.  More can be added after current release

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/IProgressReceiver.java [hunk 1]
```diff
@@ -8,10 +8,10 @@
 /**
  * An {@code IProgressReceiver} is an object that can be used to report progress on network transfers. When specified on
  * transfer operations, the {@code reportProgress} method will be called periodically with the total number of bytes
- * transferred. The user may configure this method to report progress in wha tever format desired. It is recommended
+ * transferred. The user may configure this method to report progress in whatever format desired. It is recommended
  * that this type be used in conjunction with {@link ProgressReporter#addProgressReporting(Flux, IProgressReceiver)}.
  */
-interface IProgressReceiver {
+public interface IProgressReceiver {
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/IProgressReceiver.java]: I'm not a fan of interfaces starting with `I`. I would much prefer to see this called `ProgressReceiver`.

Comment (vhvb1989) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/IProgressReceiver.java]: do we have any info about this in out guidance?  for the preferred prefix-naming stuff?  

I see a few interfaces doing the `I..` naming within the project
image

Might be a good idea to have a docs with any agreement on this.

I don't really have any specific preference regarding to this.  I am fine updating it :)

Comment (JonathanGiles) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/IProgressReceiver.java]: Yes, I will update spec to say to not do this.

Comment (JonathanGiles) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/IProgressReceiver.java]: I don't think there is any public API with the I prefix. If you do find any, please let me know.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ProgressReporter.java [hunk 2]
```diff
@@ -146,8 +146,22 @@ public void rewindProgress() {
...
+     * @param data The data whose transfer progress is to be tracked.
+     * @param progressReceiver {@link ProgressReceiver}
+     * @param lock This lock will be instantiated by the operation initiating the whole transfer to coordinate each
+     * ProgressReporterImpl.
+     * @param totalProgress We need an AtomicLong to be able to update the value referenced. Because we are already
+     * synchronizing with the lock, we don't incur any additional performance hit here by the synchronization.
+     * @return A {@code Flux} that emits the same data as the source but calls a callback to report the total amount of
+     * data emitted so far.
+     */
+    public static Flux<ByteBuffer> addParallelProgressReporting(Flux<ByteBuffer> data,
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ProgressReporter.java]: I don't think we want this method public. It's really for our own use when we handle parallel uploads

Comment (vhvb1989) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ProgressReporter.java]: we are using from `com.azure.storage.blob.specialized` (BlobAsyncClientBase#download) and from `com.azure.storage.blob` (BlobAsyncClient:upload).

We would need to make some extra refactoring to have `BlobAsyncClientBase` in same package

**Pull request merged**